### PR TITLE
Add DataBlockIndexType option in BlockBasedTableOptions

### PR DESCRIPTION
**Pull request by fgwu**
Summary:
Added DataBlockIndexType option in BlockBasedTableOptions. 
```
enum DataBlockIndexType : char {
    kDataBlockBinarySearch = 0, // traditional block type
    kDataBlockHashIndex = 1, // additional hash index appended to the end.
};
```
The default type is the traditional binary seek option: `kDataBlockBinarySearch`.

Test Plan:
unit_test: `options_settable_test`
`make -j32`

### CODE REVIEW

File: include/rocksdb/table.h [hunk 1]
```diff
@@ -100,6 +100,15 @@ struct BlockBasedTableOptions {
 
   IndexType index_type = kBinarySearch;
 
+  // The format type that will be used for the data block.
+  // Now two BlockFormatType supported
+  enum DataBlockIndexType : char {
+    kDataBlockBinarySearch = 0,  // traditional block type
+    kDataBlockHashIndex = 1,     // additional hash index appended to the end.
```
Comment (sagar0) [inline include/rocksdb/table.h]: Shall we rename this to kDataBlockHashSearch, to be similar to table IndexType ?

File: include/rocksdb/table.h [hunk 2]
```diff
@@ -100,6 +100,15 @@ struct BlockBasedTableOptions {
 
   IndexType index_type = kBinarySearch;
 
+  // The format type that will be used for the data block.
```
Comment (sagar0) [inline include/rocksdb/table.h]: s/format/index  ?

File: include/rocksdb/table.h (hunk 1 above)
Comment (fgwu) [reply include/rocksdb/table.h]: How about rename `kDataBlockBinarySearch` to be `kDataBlockNoIndex`?
I struggled a little here to find a good name. As we are naming `DataBlockIndexType`, I tried to end with -Index.
BinarySearch has no index, so the index type may be called `kDataBlockNoIndex`. But maybe `kDataBlockBinarySearch` provides more information. What do you think?
Or, we can rename the enum type as `DataBlockSearchType`. But `IndexType` seems to be more informative.

Comment (fgwu) [reply include/rocksdb/table.h]: Oh I see the existing code use `kHashSearch` etc. in `IndexType`. So never mind, I will `s/kDataBlockHashIndex/kDataBlockHashIndex/`

Comment (sagar0) [reply include/rocksdb/table.h]: Yeah, thanks. I was just trying to see if we can name it to something similar to what users are already seeing as part of our public api. 
It should be ok to keep using hash index as the name for internal implementation.

File: include/rocksdb/table.h [hunk 3]
```diff
@@ -100,6 +100,15 @@ struct BlockBasedTableOptions {
 
   IndexType index_type = kBinarySearch;
 
+  // The index type that will be used for the data block.
+  // Now two DataBlockIndexType supported
+  enum DataBlockIndexType : char {
+    kDataBlockBinarySearch = 0,  // traditional block type
+    kDataBlockHashSearch = 1,     // additional hash index appended to the end.
```
Comment (siying) [inline include/rocksdb/table.h]: Did you add an option that does nothing? Can you at least mention it in comment that it is not yet implemented?

Comment (fgwu) [reply include/rocksdb/table.h]: Just addressed in a new PR https://github.com/facebook/rocksdb/pull/4203

**Pull request closed without merging**